AIRFLOW-128 Optimize and refactor process_dag

bolkedebruin opened the pull request:
This addresses _AIRFLOW-128_.

@aoen @artwr @mistercrunch @r39132 @jlowin : ready for review.

**Goals:**
- Improve readability of the code and generic assumptions (getters should not change a state) for DagRuns
- Improve robustness and lower risk of race conditions in the scheduler
- Reduce amount of calls to the database, limit connections in the scheduler
- Identify speed optimizations possibilities

**What has changed:**
- Two new TaskInstance states have been introduced. "REMOVED" and "SCHEDULED". REMOVED will be set when taskinstances are encountered that do no exist anymore in the DAG. This happens when a DAG is changed (ie. a new version). The "REMOVED" state exists for lineage purposes. "SCHEDULED" is used when a Task that did not have a state before is sent to the executor. It is used by both the scheduler and backfills. This state almost removes the race condition that exists if using multiple schedulers: due to the fact UP_FOR_RETRY is being managed by the TaskInstance (I think that is the wrong place) is still exists for that state.
- get_active_runs was a getter that was also updating to the database. This patch refactors get_active_runs into two different functions that are now part of DagRun. 1) update_state updates the state of the dagrun based on the taskinstances of the dagrun. 2) verify_integrity checks and updates the dag run based on if the dag contains new or missing tasks. 
- DagRun.update_state has been updated to not call the database twice for the same functions. This reduces the time spent here by 50% in certain occasions when having many tasks in a Dag that need to be evaluated. Still this needs to be faster: for those Dags with many tasks the aggregation query in TaskInstance.are_dependencies_met is very expensive. It should be refactored.
- process_dag has been updated to use the functions and cleaned up, making it much more readable. Tasks are now properly locked by the database. I have played with multiprocessing here (on dagruns and taskinstances) but left it out for now. Fixing the above will help more I think.

**Stats:**
- New scheduler time spent at earlier stages is a bit more, due to eager creation of TaskIntances
- Old scheduler MAX is higher due to "are_dependencies_met()" called twice
- New scheduler fluctuates a bit more due to database locking and "are_dependencies_met" scanning the table (needs to wait for lock). 

Old:

``` bash
2016-05-23 11:17:39,031 INFO - Loop took: 0.018685 seconds
2016-05-23 11:17:44,022 INFO - Loop took: 0.013455 seconds
2016-05-23 11:17:49,033 INFO - Loop took: 0.018868 seconds
2016-05-23 11:17:54,031 INFO - Loop took: 0.019578 seconds
2016-05-23 11:17:59,024 INFO - Loop took: 0.013051 seconds
2016-05-23 11:18:05,026 INFO - Loop took: 1.010049 seconds
2016-05-23 11:18:10,399 INFO - Loop took: 1.390761 seconds
2016-05-23 11:18:32,112 INFO - Loop took: 18.104715 seconds
2016-05-23 11:19:03,432 INFO - Loop took: 31.089109 seconds
2016-05-23 11:19:14,140 INFO - Loop took: 10.492135 seconds
2016-05-23 11:19:38,339 INFO - Loop took: 24.05553 seconds
2016-05-23 11:20:06,281 INFO - Loop took: 27.887196 seconds
2016-05-23 11:20:30,215 INFO - Loop took: 23.9155 seconds
2016-05-23 11:20:53,953 INFO - Loop took: 23.375444 seconds
2016-05-23 11:21:29,168 INFO - Loop took: 35.191994 seconds
2016-05-23 11:22:42,276 INFO - Loop took: 72.384736 seconds
2016-05-23 11:23:06,276 INFO - Loop took: 23.831495 seconds
2016-05-23 11:23:38,852 INFO - Loop took: 32.333608 seconds
```

New:

``` bash
2016-05-23 11:26:12,021 INFO - Loop took: 0.011257 seconds
2016-05-23 11:26:17,031 INFO - Loop took: 0.018259 seconds
2016-05-23 11:26:22,021 INFO - Loop took: 0.01233 seconds
2016-05-23 11:26:27,026 INFO - Loop took: 0.017952 seconds
2016-05-23 11:26:32,031 INFO - Loop took: 0.017606 seconds
2016-05-23 11:26:37,707 INFO - Loop took: 0.697367 seconds
2016-05-23 11:26:43,268 INFO - Loop took: 1.255278 seconds
2016-05-23 11:27:01,234 INFO - Loop took: 14.225399 seconds
2016-05-23 11:27:03,832 INFO - Loop took: 2.580292 seconds
2016-05-23 11:27:35,556 INFO - Loop took: 29.534056 seconds
2016-05-23 11:27:55,896 INFO - Loop took: 20.321862 seconds
2016-05-23 11:28:10,192 INFO - Loop took: 14.250471 seconds
2016-05-23 11:28:40,778 INFO - Loop took: 30.337702 seconds
2016-05-23 11:28:49,003 INFO - Loop took: 8.135393 seconds
2016-05-23 11:29:09,132 INFO - Loop took: 19.923375 seconds
2016-05-23 11:29:46,856 INFO - Loop took: 37.393256 seconds
2016-05-23 11:30:30,984 INFO - Loop took: 43.79019 seconds
2016-05-23 11:31:09,856 INFO - Loop took: 38.444254 seconds
2016-05-23 11:31:31,164 INFO - Loop took: 21.061177 seconds
2016-05-23 11:32:29,776 INFO - Loop took: 58.153763 seconds
2016-05-23 11:33:12,128 INFO - Loop took: 42.105758 seconds
2016-05-23 11:33:50,796 INFO - Loop took: 38.137385 seconds
2016-05-23 11:34:40,290 INFO - Loop took: 49.115855 seconds
2016-05-23 11:35:22,900 INFO - Loop took: 42.269646 seconds
2016-05-23 11:35:38,456 INFO - Loop took: 15.453262 seconds
```

*\* Note: unittests have been added to cover process_dag **

**Dag used for testing:**
- Please note that both schedulers this Dag does not even finish. It only finished properly with the fully implemented new scheduler (including connecting dagruns by using previous ids)

``` python
from datetime import timedelta, datetime
from airflow.models import DAG, Pool
from airflow.operators import BashOperator, SubDagOperator, DummyOperator
from airflow.executors import SequentialExecutor
import airflow


# -----------------------------------------------------------------\
# DEFINE THE POOLS
# -----------------------------------------------------------------/
session = airflow.settings.Session()
for p in ['test_pool_1', 'test_pool_2', 'test_pool_3']:
    pool = (
        session.query(Pool)
        .filter(Pool.pool == p)
        .first())
    if not pool:
        session.add(Pool(pool=p, slots=8))
        session.commit()


# -----------------------------------------------------------------\
# DEFINE THE DAG
# -----------------------------------------------------------------/

# Define the Dag Name. This must be unique.
dag_name = 'hanging_subdags_n16_sqe'

# Default args are passed to each task
default_args = {
    'owner': 'Airflow',
    'depends_on_past': False,
    'start_date': datetime(2016, 04, 10),
    'retries': 0,
    'retry_interval': timedelta(minutes=5),
    'email': ['your@email.com'],
    'email_on_failure': True,
    'email_on_retry': True,
    'wait_for_downstream': False,
}

# Create the dag object
dag = DAG(dag_name,
          default_args=default_args,
          schedule_interval='0 0 * * *'
          )

# -----------------------------------------------------------------\
# DEFINE THE TASKS
# -----------------------------------------------------------------/


def get_subdag(dag, sd_id, pool=None):
    subdag = DAG(
        dag_id='{parent_dag}.{sd_id}'.format(
            parent_dag=dag.dag_id,
            sd_id=sd_id),
        params=dag.params,
        default_args=dag.default_args,
        template_searchpath=dag.template_searchpath,
        user_defined_macros=dag.user_defined_macros,
    )

    t1 = BashOperator(
        task_id='{sd_id}_step_1'.format(
            sd_id=sd_id
        ),
        bash_command='echo "hello" && sleep 1',
        dag=subdag,
        pool=pool
    )

    t2 = BashOperator(
        task_id='{sd_id}_step_two'.format(
            sd_id=sd_id
        ),
        bash_command='echo "hello" && sleep 2',
        dag=subdag,
        pool=pool
    )

    t2.set_upstream(t1)

    sdo = SubDagOperator(
        task_id=sd_id,
        subdag=subdag,
        retries=0,
        retry_delay=timedelta(seconds=5),
        dag=dag,
        depends_on_past=False,
        executor=SequentialExecutor()
    )

    return sdo

start_task = DummyOperator(
    task_id='start',
    dag=dag
)

for n in range(1, 17):
    sd_i = get_subdag(dag=dag, sd_id='level_1_{n}'.format(n=n), pool='test_pool_1')
    sd_ii = get_subdag(dag=dag, sd_id='level_2_{n}'.format(n=n), pool='test_pool_2')
    sd_iii = get_subdag(dag=dag, sd_id='level_3_{n}'.format(n=n), pool='test_pool_3')

    sd_i.set_upstream(start_task)
    sd_ii.set_upstream(sd_i)
    sd_iii.set_upstream(sd_ii)
```

criccomini commented on airflow/models.py:
```diff
@@ -1061,6 +1060,7 @@ def are_dependencies_met(
 
         task = self.task
 
+        logging.info("Checkpoint A")
```
Assuming this is for debugging, and will go away before final merge

bolkedebruin replied:
Absolutely, everything after my initial commit in this pr is actually profiling stuff (ie. WIP). A lot of time is spent in are_dependencies_met as it is iterating over _all_ tasks now every time now.

criccomini commented on airflow/models.py:
```diff
@@ -3455,14 +3368,170 @@ def find(dag_id, run_id=None, state=None, external_trigger=None, session=None):
+
+            logging.info("Checkpoint 5")
+
+    # todo: determine we want to use with_for_update to make sure to lock the run
```
Spacing seems off here

jlowin commented on airflow/jobs.py:
```diff
@@ -484,84 +484,99 @@ def process_dag(self, dag, queue):
-                self.logger.debug('Adding task: {} to the COULD_NOT_RUN set'.format(ti))
-                could_not_run.add(ti)
+        for run in active_dag_runs:
+            tis = run.get_task_instances(session=session, state=(State.NONE,
```
I think this line is too restrictive.

As an example, using CeleryExecutor but no workers, this DAG is almost immediately marked as failed (I had it on hand from testing a different issue):
https://gist.github.com/jlowin/4f0c88b80f35ac45862c6222d153abcf

I think what happens is that the first set of tasks are loaded and sent to the Executor, so their states are set to RUNNING even though there are no celery workers. The next time through the loop, the scheduler gets to this line and loads up the one remaining task -- one that is also downstream from all the RUNNING tasks and therefore can't run yet. It evaluates it, finds it can't run, and says the DAG is deadlocked because `could_not_run == available_tasks`. 

So I think we need to account for the fact that tasks may be running or in transit to the executor before moving on to deadlock assumption.

bolkedebruin replied:
As discussed on gitter. The check for deadlock here does not make sense anymore. TaskInstances are sent to the executor in a "SCHEDULED" state or in "UP_FOR_RETRY". The count of available tasks can therefore equal the amount of tasks that are not runnable as tasks can actually be running at that moment. "update_state" does check for deadlocked dag_runs by using all relevant states.

criccomini commented on airflow/models.py:
```diff
@@ -106,6 +106,10 @@ def clear_task_instances(tis, session, activate_dag_runs=True):
             if ti.job_id:
                 ti.state = State.SHUTDOWN
                 job_ids.append(ti.job_id)
+        # todo: this creates an issue with the webui tests
```
WIP? or is this for a future refactor PR?

bolkedebruin replied:
Future. For lineage purposes this needs to be done, now you would loose that. verify_integrity will recreate the tasks if required.

criccomini commented on airflow/jobs.py:
```diff
@@ -484,84 +484,78 @@ def process_dag(self, dag, queue):
-        db_dag.scheduler_lock = False
-        session.merge(db_dag)
-        session.commit()
+            # this is actually quite s slow loop and could
```
Nit: s slow -> a slow

criccomini commented on airflow/jobs.py:
```diff
@@ -484,84 +484,78 @@ def process_dag(self, dag, queue):
+                if task.adhoc:
+                    continue
+
+                if ti.is_premature():
```
Comment here would improve readability.

criccomini commented on airflow/jobs.py:
```diff
@@ -484,84 +484,78 @@ def process_dag(self, dag, queue):
+                # todo: ti.task is transient but needs to be set
+                ti.task = task
+                # todo: remove adhoc
+                if task.adhoc:
```
Comment here would improve readability.

jlowin commented on airflow/models.py:
```diff
@@ -3455,14 +3366,176 @@ def find(dag_id, run_id=None, state=None, external_trigger=None, session=None):
         if external_trigger:
             qry = qry.filter(DR.external_trigger == external_trigger)
-        dr = qry.all()
+        dr = qry.order_by(DR.execution_date).all()
```
Is this `order_by` necessary? I'm sure it adds overhead -- might be better to let users sort the list after calling `find()` if they need it sorted rather than sort it every time.

bolkedebruin replied:
Sorting is faster done by the db. I could add a sort by parameter. But overhead is very small as it will be on the index

jlowin commented on airflow/models.py:
```diff
@@ -3455,14 +3366,176 @@ def find(dag_id, run_id=None, state=None, external_trigger=None, session=None):
+            state=State.unfinished(),
+            session=session
+        )
+        has_unfinished_tasks = True if unfinished_tasks else False
```
It looks like this might not be needed, though maybe it makes things more clear. For example, below:

`all_deadlocked = (has_unfinished_tasks and no_dependencies_met)`
could be:
`all_deadlocked = (unfinished_tasks and no_dependencies_met)` since it will implicitly call `bool(unfinished_tasks)`

Similarly 
`elif has_unfinished_tasks and none_depends_on_past and no_dependencies_met:`
Could be
`elif unfinished_tasks and none_depends_on_past and no_dependencies_met:`

bolkedebruin replied:
Ah yes. Forgot that

jlowin commented on airflow/models.py:
```diff
@@ -3455,14 +3366,176 @@ def find(dag_id, run_id=None, state=None, external_trigger=None, session=None):
+        # dags look deadlocked incorrectly. Later we will check individual
+        # dagruns, as long as they don't have depends_on_past=True
+
+        all_deadlocked = (has_unfinished_tasks and no_dependencies_met)
```
I'm not sure this logic (`all_deadlocked`) still works here, since now the only TI's being considered are from a single DagRun.

Let's say there are two DagRuns with a `depends_on_past=True` for the first task. The first DagRun will run normally. I think the second DagRun will report that it has unfinished tasks but that they can't run due to `depends_on_past` and it will deadlock. 

I think we should stop checking for that type of Deadlock, for simplicity. The deadlock in line 3489 will check if just this DAG is deadlocked, though it will do it in a way that respects the weird behavior of depends_on_past.

jlowin commented on airflow/models.py:
```diff
@@ -3455,14 +3366,176 @@ def find(dag_id, run_id=None, state=None, external_trigger=None, session=None):
+        :return: DAG
+        """
+        if not self.dag:
+            raise AirflowException("{}.dag needs to be set".format(self))
```
nit: this is really minor but I don't think it will be clear from this error message that `XXX.dag` is referring to the DAG attribute of this DagRun. I think it will look something like `<DagRun foo bar>.dag`. Maybe saying `The DAG for {} needs to be set` would be more clear?

criccomini commented on airflow/jobs.py:
```diff
@@ -401,11 +402,14 @@ def schedule_dag(self, dag):
+                    or_(DagRun.external_trigger == False,
+                    # add % as a wildcard for the like query
+                    DagRun.run_id.like(DagRun.ID_PREFIX+'%')
+                        )
```
Spacing here seems weird

criccomini commented on airflow/jobs.py:
```diff
@@ -643,13 +640,13 @@ def prioritize_queued(self, session, executor, dagbag):
-        This function splits a list of dags into chunks of int size.
-        _split_dags([1,2,3,4,5,6], 3) becomes [[1,2,3],[4,5,6]]
+        This function splits a list of items into chunks of int size.
+        _split_items([1,2,3,4,5,6], 3) becomes [[1,2,3],[4,5,6]]
```
_split, not _split_items, right?

bolkedebruin replied:
Ah yes. Will fix that

criccomini commented on airflow/jobs.py:
```diff
@@ -700,6 +697,7 @@ def _execute(self):
 
                 if len(self.dag_ids) > 0:
                     dags = [dag for dag in dagbag.dags.values() if dag.dag_id in self.dag_ids]
+                    self.logger.info("Dags: {}".format(dags))
```
How frezuently does _do_dags get run? Concerned this might be a bit verbose.

bolkedebruin replied:
I think that is debugging info. I will change it to debug.

bolkedebruin committed:
Optimize and refactor process_dag

This patch addresses the following issues:
get_active_runs was a getter that was also updating
to the database. This patch refactors get_active_runs
into two different functions that are part of DagRun.
update_state update state of the dagrun based on the
taskinstances of the dagrun. verify_integrity checks
and updates the dag run based on if the dag contains
new or missing tasks.

Deadlock detection has been refactored to ensure that
database does not get hit twice, in some circumstances
this can reduce the time spent by 50%.

process_dag has been refactored to use the functions
of DagRun reducing complexity and reducing pressure on the
database. In addition locking is now properly working
under the assumption that the heartrate is longer
than the time process_dag spends.

Two new TaskInstance states have been introduced. "REMOVED"
and "SCHEDULED". REMOVED will be set when taskinstances
are encountered that do no exist anymore in the DAG.
This happens when a DAG is changed (ie. a new version).
The "REMOVED" state exists for lineage purposes.
"SCHEDULED" is used when a Task that did not have a state
before is sent to the executor. It is used by both the
scheduler and backfills. This state almost removes the
race condition that exists if using multiple schedulers:
due to the fact UP_FOR_RETRY is being managed by the
TaskInstance (I think that is the wrong place) is still
exists for that state.

mistercrunch commented on airflow/jobs.py:
```diff
@@ -401,11 +402,14 @@ def schedule_dag(self, dag):
-                        or_(DagRun.external_trigger == False,
-                            # add % as a wildcard for the like query
-                            DagRun.run_id.like(DagRun.ID_PREFIX+'%')))
+                dag_id=dag.dag_id).filter(
```
my new preferred way to indent method chains is 

```
qry = (
    session.query(func.max(DagRun.execution_date))
    .filter_by(dag_id = dag.dag_id)
    .filter(or_(
        DagRun.external_trigger == False,
        DagRun.run_id.like(DagRun.ID_PREFIX+'%')
    )
)
```

plypaul commented on airflow/models.py:
```diff
@@ -3438,36 +3346,186 @@ def refresh_from_db(self, session=None):
+            state=State.unfinished(),
+            session=session
+        )
+        none_depends_on_past = all(t.task.depends_on_past for t in unfinished_tasks)
```
Should this be `not t.task.depends_on_past`?
